ptt.php

```diff
@@ -1,156 +1,179 @@
-    echo "<URLDown>RTPTx:Stop</URLDown>";
-    echo "</SoftKeyItem>";
+
+class Push2Talk {
+	const XML_EXECUTE 		= '<CiscoIPPhoneExecute>%s</CiscoIPPhoneExecute>';
+	const XML_EXECUTE_ITEM 	= '<ExecuteItem Priority="%d" URL="%s"/>';
+	const URI_START			= 'RTPMRx:%s:%d';
+
+	public $multicastAddress= '225.3.15.13';
+	public $multicastPort 	= 20480;
+
+	var $authName 			= 'cisco';
+	var $authPassword 		= 'cisco';
+	var $deviceName 		= null;
+
+	var $devices 			= array();
+
+	function __construct(){ }
+
+	function setAuthentication($authName, $authPassword){
+		$this->authName 	= $authName;
+		$this->authPassword = $authPassword;
+	}
+
+	function addDevice($deviceIp){
+		$this->devices[] 	= new Device($deviceIp, $this->authName, $this->authPassword);
+	}
+
+	function addDevices(array $ips){
+		foreach($ips as $ip){
+			$this->addDevice($ip);
+		}
+	}
+
+	function start(){
+		$this->execute(array(
+			sprintf(Push2Talk::URI_START, $this->multicastAddress, $this->multicastPort)
+		));
+	}
+
+	function stop(){
+		$this->execute(array('RTPRx:Stop','RTPTx:Stop','Init:Services'));
+	}
+
+	function execute(array $uris, $priority = 0){
+		$xmlExecute = array_map(function ($value) {
+			return sprintf(Push2Talk::XML_EXECUTE_ITEM, $priority, $value);
+		}, $uris); 															// build ExecuteItem xml data-array
+		$xml = sprintf(Push2Talk::XML_EXECUTE, implode('',$xmlExecute));	// build xml data including ExecuteItem
+
+		foreach($this->devices as $device) {
+			$device->push($xml);
+		}
+	}
 }
 
-echo "</CiscoIPPhoneText>";
-?>
+
+$push2Talk = new Push2Talk;
+$push2Talk->setAuthentication('cisco', 'cisco');
+$push2Talk->addDevices(array('10.0.2.225', '10.0.2.227'));
+
+// optional settings
+//$push2Talk->multicastAddress = '225.3.15.13';
+//$push2Talk->multicastPort = 16384
+
+
+$response = '';
+do {  // while loop for error handling
+
+	$deviceName =  null;
+	$isTalking = false;
+	$action = isset($_GET['action']) ? $_GET['action'] : 'start';
+
+	if (!isset($_GET['name'])){
+		$response = '<CiscoIPPhoneText><Title>Error!</Title><Text>No MAC provided by phone!</Text></CiscoIPPhoneText>';
+		break;
+	}
+
+	$deviceName = $_GET['name'];
+
+	switch($action){
+		default:
+			$push2Talk->start();
+			break;
+
+		case 'close':
+			$push2Talk->stop();
+			break;
+	}
+
+	// Cisco headers
+	$response = '<CiscoIPPhoneText><Title>PTT</Title><Prompt>Use soft keys to talk</Prompt>';
+
+	// PTT operation
+	if (!$isTalking) {
+		$response .= "<Text>Press and hold the [Talk] soft key to transmit. If you want to speak continuously without holding a button down, you can press and release the [Lock] soft key.</Text>";
+
+		$response .=  "<SoftKeyItem>";
+		$response .=  "<Name>Talk</Name>";
+		$response .=  "<URL>RTPTx:Stop</URL>";
+		$response .=  "<Position>1</Position>";
+		$response .=  "<URLDown>RTPMTx:{$push2Talk->multicastAddress}:{$push2Talk->multicastPort}</URLDown>";
+		$response .=  "</SoftKeyItem>";
+
+		$response .=  "<SoftKeyItem>";
```

Comment:
initiator:Exit -> stop and close page on all participants
participant:Exit -> close page only, take device out of devicesArray, to prevent further updates.